src/constructRoutes.js

```diff
@@ -43,10 +47,126 @@ import {
  * @returns {ResolvedRoutesConfig}
  */
 export function constructRoutes(routesConfig) {
+  if (routesConfig && routesConfig.nodeName) {
+    routesConfig = domToRoutesConfig(routesConfig);
+  }
+
   validateAndSanitize(routesConfig);
   return routesConfig;
 }
 
+/**
+ * Converts a domElement to a json object routes config
+ *
+ * @param {HTMLElement} domElement
+ * @returns {InputRoutesConfigObject}
+ */
+function domToRoutesConfig(domElement) {
+  if (domElement.nodeName.toLowerCase() !== "single-spa-router") {
+    throw Error(
+      `single-spa-layout: The HTMLElement passed to constructRoutes must be <single-spa-router>. Received ${domElement.nodeName}`
+    );
+  }
+
+  const result = {
+    sourceElement: domElement,
+    routes: [],
+  };
+
+  if (getAttribute(domElement, "mode")) {
+    result.mode = getAttribute(domElement, "mode");
+  }
+
+  if (getAttribute(domElement, "base")) {
+    result.base = getAttribute(domElement, "base");
+  }
+
+  for (let i = 0; i < domElement.childNodes.length; i++) {
+    result.routes.push(
+      ...elementToJson(domElement.childNodes[i], domElement.parentNode)
+    );
+  }
+
+  return result;
+}
+
+function getAttribute(element, attrName) {
+  if (inBrowser) {
+    // browser
+    return element.getAttribute(attrName);
+  } else {
+    // NodeJS with parse5
+    const attr = element.attrs.find((attr) => attr.name === attrName);
+    return attr ? attr.value : null;
+  }
+}
+
+/**
+ * @param {HTMLElement} element
+ * @param {HTMLElement} parentElement
+ * @returns {Array<Route>}
+ */
+function elementToJson(element, parentElement) {
+  if (element.nodeName.toLowerCase() === "application") {
+    if (element.childNodes.length > 0) {
+      throw Error(
+        `<application> elements must not have childNodes. You must put in a closing </application> - self closing is not allowed`
+      );
+    }
+    const application = {
+      type: "application",
+      containerEl: parentElement,
```

Comment:
This part will change to domElementGetter stuff, but that will be a separate refactor. For now, I think that this is okay to merge in, since it's part of the current schema for resolvedroutesconfig. I'll remove it in a future PR